File: Self-Study/쿼드트리_1992.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class ÄõµåÆ®¸®_1992 {
	static int N;
	static int[][] img;
	static StringBuilder sb;
	public static void main(String[] args) throws NumberFormatException, IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		N = Integer.parseInt(br.readLine());
		img = new int[N][N];
		for(int i=0; i<N; i++) {
			String str = br.readLine();
			for(int j=0; j<N; j++) {
				img[i][j] = str.charAt(j) - '0';
			}
		}
		sb = new StringBuilder();
		sb.append("(");
		dfs(0, 0, N);
		
	}
	private static StringBuilder dfs(int r, int c, int n) {
		if(n == 1) {
			sb.append(img[r][c]);
			return sb;
		}
		
		StringBuilder d1 = dfs(r, c, n/2);
		if(d1.equals("0000"))
				d1.
		StringBuilder d2 = dfs(r, c+n/2, n/2);
		StringBuilder d3 = dfs(r+n/2, c, n/2);
		StringBuilder d4 = dfs(r+n/2, c+n/2, n/2);
		
		
		
		return d4;
		
		
	}

}

File: Self-Study/쿼드트리_1992.py
n = int(input())
img = [list(input()) for _ in range(n)]


def dfs(r, c, d):
    if d == 1:
        return img[r][c]
    d //= 2
    a = ''  #깊이마다 문자열을 저장

    a += dfs(r, c, d)           #왼쪽위
    a += dfs(r, c + d, d)       #오른쪽위
    a += dfs(r + d, c, d)       #왼쪽아래
    a += dfs(r + d, c + d, d)   #오른쪽아래

    if a == '0000':
        return '0'
    elif a == '1111':
        return '1'
    else:
        return '(' + a + ')'


print(dfs(0, 0, n))
